src/librustc_resolve/lib.rs

```diff
@@ -2283,24 +2283,25 @@ impl<'a> Resolver<'a> {
                 PatKind::Ident(bmode, ref ident, ref opt_pat) => {
                     // First try to resolve the identifier as some existing
                     // entity, then fall back to a fresh binding.
-                    let resolution = if let Ok(resolution) = self.resolve_path(pat.id,
-                                &Path::from_ident(ident.span, ident.node), 0, ValueNS) {
+                    let local_def = self.resolve_identifier(ident.node, ValueNS, true);
```

Comment:
Why is this preferable?
I used `resolve_path` solely for stylistic reasons - it returns exactly what we want here (`PathResolution`), and returns "fully correct" result with all the adjustments like `Def::Local` -> `Def::Upvar` (even if they don't matter here).
I also don't like mixing "intermediate" resolution byproducts like `LocalDef` with final `PathResolution`s, again for stylistic reasons - high-level routines like `resolve_pattern` should work only with the latter. Can `resolve_identifier` return fully adjusted `Result<PathResolution>`, like `resolve_path`?